### ISSUE TITLE
Integrate Thunderbolt 3 Patch from pierpaolodimarzo

### ISSUE THREAD

**Issue by Pete466** [valnoxy/t480-oc#164]
I want to ask whether the Thunderbolt 3 Patch from here: https://github.com/pierpaolodimarzo/ThinkPad-T480
can be integrated into the EFI. On my system this patch works on Sequoia and I can provide my EFI if there is an interest in that.

**Comment by galaxy-nomad** [comment]
I just bought one and wondering if I can get this working. I'm using MacOS Ventura if that makes any difference. Are you using the Thuderbolt 3 patch with a T480?

**Comment by Pete466** [comment]
yes sir, I can provide the EFI. The crucial thing is, that the Thunderbolt-Firmware is updated to the most recent version. That's also due to the fact, that else the Thunderbolt-Port will eventually break. That's a design flaw by Lenovo. Once you have done that, I can upload my EFI and then it will work in principle, just that it may needs to be optimized (I have no deeper knowledge to understand what makes it actually work or not). The Thunderbolt-Port is the one facing you btw., the other one won't work

**Comment by galaxy-nomad** [comment]
Great please upload and thank you!

**Comment by Pete466** [comment]
Here you go: https://anonymfile.com/Y1kAR/efi.zip
Hope it works for you and if you happen to be able to improve on it, would love to hear from you!

**Comment by galaxy-nomad** [comment]
Thanks!

**Comment by galaxy-nomad** [comment]
@Pete466 Do you have a stable build? Im running macos ventura and having loads of issues. My external monitor will not boot or wake from sleep without being switched on or off and on while the laptop is open. The whole system will crash out of nowhere and the computer just shut off like someone pulled the plug. Also the track pad will not respond when waking the laptop. It seems super unstable and unpredictable.

**Comment by Pete466** [comment]
I haven't been running Ventura on it, but Sonoma and Sequoia. It's semi-stable, as in no crashes, trackpad works fine as long as you don't put it into sleep. I never put the laptop into sleep because that causes - as you are saying - lots of issues. For example the laptop will randomly wake up from sleep and the Thunderbolt will stop working. Thunderbolt in my implementation works, if you start the laptop with the monitor being on. As soon as you disconnect it however, it fails to reconnect etc. For me it's a "better than no Thunderbolt at all" kind of solution and I don't have the knowledge to do it better. Maybe @valnoxy can help us out on this one

**Comment by lolipuru** [comment]
![image](https://github.com/user-attachments/assets/69c968ec-ab80-4fe1-8d13-9282aec9ac1c)

+1 for pierpaolodimarzo's thunderbolt patch, but iirc hotplug isn't working, i might prolly trying thunderbolt patches from https://github.com/tylernguyen/x1c6-hackintosh (for proper hotplug and 3.1 support) later when i have time

**Comment by galaxy-nomad** [comment]
UPDATE: I switched to the non-touch EFI-i5.8250u.zip build from https://github.com/pierpaolodimarzo/ThinkPad-T480

I have a T480 i7 8650U with 64GB RAM 

So far everything seems to be working and Thunderbolt is operating. I don't have any thunderbolt peripherals to test but my usb-c external hard drive mounts when connecting to the thunderbolt port. I had to do the FW 23 update from a windows boot drive to get macos to even see the Thunderbolt port. 

My only issue now is external HDMI monitor will not wake from sleep and you cannot boot in clamshell mode as external monitor will not be activated without turning it on or plugging it in with the laptop lid open.

**Comment by Pete466** [comment]
yea, if you remain on Ventura that makes the most sense. I wanted to use recent versions, thus the attempt to update it. @lolipuru if you have success, please share it with us, that'd be awesome

**Comment by vanillaogsky** [comment]
> I want to ask whether the Thunderbolt 3 Patch from here: https://github.com/pierpaolodimarzo/ThinkPad-T480 can be integrated into the EFI. On my system this patch works on Sequoia and I can provide my EFI if there is an interest in that.

are you using this EFI? is this working without heliport?

**Comment by Pete466** [comment]
I made my own, see the link, by trying to mix them with limited understanding. The Thinkpad was running on Sequoia and now on Sonoma, because I couldn't get Heliport to work at all (app refused to start, currently I use airportwltm). Thunderbolt worked, with the limitation of sleep / disconnect and reconnect then it would stop to work till reboot and cause temporary boot issues

**Comment by lolipuru** [comment]
> I made my own, see the link, by trying to mix them with limited understanding. The Thinkpad was running on Sequoia and now on Sonoma, because I couldn't get Heliport to work at all (app refused to start, currently I use airportwltm). Thunderbolt worked, with the limitation of sleep / disconnect and reconnect then it would stop to work till reboot and cause temporary boot issues

technically, you can get airportltwm running on sequoia by spoofing intel wifi card to boardcom, then run oclp patch 

> UPDATE: I switched to the non-touch EFI-i5.8250u.zip build from https://github.com/pierpaolodimarzo/ThinkPad-T480
> 
> I have a T480 i7 8650U with 64GB RAM
> 
> So far everything seems to be working and Thunderbolt is operating. I don't have any thunderbolt peripherals to test but my usb-c external hard drive mounts when connecting to the thunderbolt port. I had to do the FW 23 update from a windows boot drive to get macos to even see the Thunderbolt port.
> 
> My only issue now is external HDMI monitor will not wake from sleep and you cannot boot in clamshell mode as external monitor will not be activated without turning it on or plugging it in with the laptop lid open.

unfortunately, my hdmi monitor is dead. so i don't have any changes to try fix it properly yet.

**Comment by lolipuru** [comment]
![image](https://github.com/user-attachments/assets/8ddf53c2-b188-4a96-b6b7-6f071db8e85f)

a pretty funny thing when i replaced with x1c6 hotplug patch is, thunderbolt port now return as expresscard instead....

i think i might need to fix a little bit for TB to be properly loaded in PCI (i don't have any thunderbolt devices support, so i can't properly test it)

**Comment by Pete466** [comment]
I could assist with the testing bit, as the laptop is currently running macOS still. Regarding Airport you're right, I saw that earlier, but haven't found the time to try implementing it...

**Comment by silikonz** [comment]
> Here you go: https://anonymfile.com/Y1kAR/efi.zip Hope it works for you and if you happen to be able to improve on it, would love to hear from you!

Hi @Pete466, the link seems to be dead... would you be able to upload it again? Thanks :)

**Comment by Pete466** [comment]
Here you go: https://anonymfile.com/ad2Qe/efi.zip
